karabo_bridge/server.py

```diff
@@ -0,0 +1,252 @@
+            self.server_socket = self.zmq_context.socket(zmq.PUSH)
+        else:
+            raise ValueError(f'Unsupported socket type: {sock}')
+        self.server_socket.setsockopt(zmq.LINGER, 0)
+        self.server_socket.set_hwm(1)
+        self.server_socket.bind(endpoint)
+
+        self.stopper_r = self.zmq_context.socket(zmq.PAIR)
+        self.stopper_r.bind('inproc://sim-server-stop')
+        self.stopper_w = self.zmq_context.socket(zmq.PAIR)
+        self.stopper_w.connect('inproc://sim-server-stop')
+
+        self.poller = zmq.Poller()
+        self.poller.register(self.server_socket, zmq.POLLIN | zmq.POLLOUT)
+        self.poller.register(self.stopper_r, zmq.POLLIN)
+
+    @property
+    def endpoint(self):
+        endpoint = self.server_socket.getsockopt_string(zmq.LAST_ENDPOINT)
+        endpoint = endpoint.replace('0.0.0.0', gethostname())
+        return endpoint
+
+    def send(self, data, metadata=None):
+        payload = self.dump(data, metadata)
+        events = dict(self.poller.poll())
+
+        if self.stopper_r in events:
+            self.stopper_r.recv()
+            return True
+
+        if events[self.server_socket] is zmq.POLLIN:
+            msg = self.server_socket.recv()
+            if msg != b'next':
+                print(f'Unrecognised request: {msg}')
+                self.server_socket.send(b'Error: bad request %b' % msg)
+                return
+
+        self.server_socket.send_multipart(payload, copy=False)
+
+
+class SimServer(Sender):
+    def __init__(self, endpoint, sock='REP', ser='msgpack',
+                 protocol_version='2.2', detector='AGIPD', raw=False,
+                 nsources=1, datagen='random', data_like='online', *,
+                 debug=True):
+        super().__init__(endpoint, sock=sock, protocol_version=protocol_version)
+
+        if ser != 'msgpack':
+            raise ValueError("Unknown serialisation format %s" % ser)
+
+        self.data = DataGenerator(
+            detector=detector, raw=raw, nsources=nsources, datagen=datagen,
+            data_like=data_like, debug=debug)
+        self.debug = debug
+
+    def loop(self):
+        print(f'Simulated Karabo-bridge server started on:\n'
+              f'tcp://{self.endpoint}')
+
+        timing_interval = 50
+        t_prev = time()
+        n = 0
+
+        for data, meta in self.data:
+            done = self.send(data, meta)
+            if done:
+                break
+
+            if self.debug:
+                print('Server : emitted train:',
+                      next(v for v in meta.values())['timestamp.tid'])
+            n += 1
+            if n % timing_interval == 0:
+                t_now = time()
+                print('Sent {} trains in {:.2f} seconds ({:.2f} Hz)'
+                      ''.format(timing_interval, t_now - t_prev,
+                                timing_interval / (t_now - t_prev)))
+                t_prev = t_now
+
+
+class ServerInThread(Sender, Thread):
```

Comment:
I'm always wary of multiple inheritance; I think it makes it harder to read code, because methods and attributes can come from two different parent classes.

In this case, I think it's quite easy to use composition, so it inherits from `Sender` and has a `Thread` as an attribute:

```python
self.thread = Thread(target=self._run)
```